Add keyfile option to keepassxc cli import cmd

Colfenor opened the pull request:
Fixes #5311

Basically added the add keyFile logic from the create
command to the import command and moved the `loadFileKey()` function
to the `Utils` class since it is now used in both create & import classes.


## Testing strategy
export an existing Database with a keyfile to .xml

import it again using the keyfile option in the import command: (e.g.)
`./keepassxc-cli import -k ~/test.key ~/test.xml ~/test.kdbx`

at the current implementation state the filePath parameters have to be in that exact order.

## Type of change
- ✅ New feature (change that adds functionality)

louib commented:
Hi @Colfenor  ! Thanks for the PR :pray: 

There is a manpage for the CLI located in `docs/man/` which would need to be updated with the new option.

Tests for the CLI are located in `tests/TestCli.cpp` and I think it's definitely worth it to add tests for the new option. For the CLI tests to run, you need to configure the project with `-DWITH_GUI_TESTS=ON`. You probably can use the test for the `create` command as a template.

Also, there seems to be some formatting issues with the PR. Formatting can be performed automatically by calling make format from the `build/` directory. See https://github.com/keepassxreboot/keepassxc/blob/develop/.github/CONTRIBUTING.md#coding-styleguide for details.

Colfenor commented:
Hi @louib ! Thanks a lot for your info on the `make format` option and on the gui test option.
I added the man-page documentation, fixed format issues and added import cmd tests.

louib commented on src/cli/Import.cpp:
```diff
@@ -40,12 +41,20 @@
  * @return EXIT_SUCCESS on success, or EXIT_FAILURE on failure
  */
+
+const QCommandLineOption Import::SetKeyFileOption =
```
@Colfenor I think we should reuse the option as defined in `Create`, since they are exactly the same. Also in the future I believe we should centralize the database creation options and behavior, so that would be a step in that direction.

Colfenor replied:
jup, I changed it to using the `Create`. Dodging some code duplication :)

louib reviewed (commented):
The password setting behavior is not the same for the `db-create` and `import` commands. By default, we don't ask for a password when creating the db, but we do when importing. I feel like the behavior should be the same, whichever we choose. @droidmonkey what do you think?

droidmonkey commented:
The behavior should be the same because its the same action "create a db" just one is blank the other has data in it.

Colfenor commented:
> The behavior should be the same because its the same action "create a db" just one is blank the other has data in it.

okidok, I added the SetPasswordOption to the import command and updated man-page + tests. 
Moreover the `db-create` and `import` cmd only accept `-p -k`in this exact order. I feel like this should be further clarified in the man pages, that exactly this order is accepted.

louib commented:
> Moreover the db-create and import cmd only accept -p -k in this exact order.

Is it not because the `-k` option takes an argument and not the `-p` option? I feel like `-k key_file_path.txt -p` should work, no?

Colfenor commented:
> 
> 
> > Moreover the db-create and import cmd only accept -p -k in this exact order.
> 
> Is it not because the `-k` option takes an argument and not the `-p` option? I feel like `-k key_file_path.txt -p` should work, no?

Ah yes, I missed the argument. I just tested your statement, and can confirm that `-k key_file_path.txt -p` works :)

louib commented on docs/man/keepassxc-cli.1.adoc:
```diff
@@ -251,6 +256,13 @@ The same password generation options as documented for the generate command can
+  Set the key file for the database.
+
+*-p*, *--set-password*::
+  Set a password for the database.
```
@Colfenor at that point, the only difference between the Create options and the Import options is that we allow to set the decryption time in `db-create`. How hard would it be to reuse the same database creation code for `db-create` and `import`? This way we would make sure we always support the same options in both commands, and we could also merge the documentation in the manpage under a `Create and Import options` header, like we do for other commands that same options

Colfenor replied:
Hi @louib ! Yes I totally agree, to provide a consistent behavior of both commands I added the code of the decryption time parameter option to the import command. I tested it locally and it works. Moreover the manpage is now updated aswell as the cli-tests. Ready for review :)

louib replied:
The code for the database creation is duplicated in `Import` and `Create`, no? Could we create a static function in `Create` that takes the args parser and return a db, and use that function in both `Create` and `Import`?

Is `db.setKdf(KeePass2::uuidToKdf(KeePass2::KDF_ARGON2));` what we do when a db is created from the UI?

Colfenor replied:
> The code for the database creation is duplicated in `Import` and `Create`, no? Could we create a static function in `Create` that takes the args parser and return a db, and use that function in both `Create` and `Import`?

The first part of the database creation is duplicated exactly however, later on in the **Create** cmd the db object is initialized as `QSharedPointer<Database> db(new Database);` on the heap whereas, in the **Import** cmd we have a simple `Database db;` stack object which gets operated on.

I'm not entirely sure on why there is a difference, both ways seem to work so far from the cli.
However maybe we have a bug here ?
 
> Is `db.setKdf(KeePass2::uuidToKdf(KeePass2::KDF_ARGON2));` what we do when a db is created from the UI?

Yes exactly :) 

https://github.com/keepassxreboot/keepassxc/blob/develop/src/gui/dbsettings/DatabaseSettingsWidgetEncryption.cpp#L88

louib replied:
@Colfenor 

If we're defaulting to ARGON2 in the GUI I guess we should be doing the same in the CLI. We could also created a `DEFAULT_KDF` constant available to both and use that.

> The first part of the database creation is duplicated exactly, however later on in the Create cmd the db object is initialized as QSharedPointer<Database> db(new Database); on the heap whereas, in the Import cmd we have a simple Database db; stack object which gets operated on.

I feel like both should be handled exactly the same. Not sure which one is best tho. If both work right now I'd say pick one, and if we change in the future we'll only have to change it in the static function.

Colfenor replied:
@louib 

> If we're defaulting to ARGON2 in the GUI I guess we should be doing the same in the CLI. 
Ok, changed it.

> I feel like both should be handled exactly the same. Not sure which one is best tho. If both work right now I'd say pick one, and if we change in the future we'll only have to change it in the static function.

Alright, I chose to create the db on the stack. The cli-tests still work and I tested it locally.

louib commented on src/cli/Create.cpp:
```diff
@@ -70,9 +71,8 @@ Create::Create()
  * @return EXIT_SUCCESS on success, or EXIT_FAILURE on failure
  */
-int Create::execute(const QStringList& arguments)
+int Create::createDataBase(const QSharedPointer<QCommandLineParser>& parser, bool importDatabase)
```
@Colfenor I feel like we could be able to remove the `importDatabase` argument. We could just create an empty database in the case of a `Create`, and import the xml in the case of `Import`, and then pass to db to this function, which would set the keys. The function would be named `setEncryptionKeys` though. What do you think?

Colfenor replied:
I would say it basically breaks down of how the initializing of `QString databaseFilename` is different in the 2 commands. In Create it is the first parsed argument vs in Import it is the second argument because the first one would be the xml filePath. 

Now the `setEncryptionKeys` function would need then the `databaseFilename` string and the database object as arguments both are then needed in the final `dataBase.saveAs(.....)` attempt :)

louib replied:
The other option would be to save the database after calling `setEncryptionKeys`. This would lead to a bit of code duplication, but we have that database saving code in a lot of the CLI commands anyway. At least all of the code related to handling the encryption options would be centralized.

Colfenor replied:
Hm I checked on this option, it's in my opinion not ideal. 
Just to make sure we are on the same page, would your suggested function `setEncryptionKeys` return a `QSharedPointer<CompositeKey>` ?

If so we still have the code of the `decryptionTime` duplicated which is determined from the parser object and later on set on the actual new database object, which again would be outside of `setEncryptionKeys` in this design choice. 

But yeah the **keyfile & password** setting code would be centralized.

louib replied:
> would your suggested function setEncryptionKeys return a QSharedPointer<CompositeKey>

It could return a `QSharedPointer<Database>` instead? This way we could set the decryption time also in the function. I guess `setEncryptionKeys` is not a proper name. In the future, we would also handle options like setting the database name and the database description in that function, for example. Maybe `createDatabaseFromOptions`?

Colfenor replied:
> It could return a `QSharedPointer<Database>` instead? 

Yeah sure :) similar to the `unlockDatabase` function in `Utils` ! So for me the question still remains what should happen in the Create or Import `::execute` function. Currently from the latest commit it calls `createDatabase`where all the logic is contained.

If some logic like the `databaseFilename` creation and finally receiving the `db` object + checking/saving it, should be for instance in `::execute` we still have the parser duplicated which we would need in both functions.

louib replied:
Not sure what you mean by `have the parser variable duplicated`, but both commands would access the parser variable in their `execute` function for saving the database. I think this is inevitable. I don't see a problem with both the execute functions and the `createDatabaseFromOptions` handling the `parser` variable.

Maybe we could rename it `initializeDatabaseFromOptions`, since the actual creation (i.e. saving or importing) will be done after calling that function, in the `execute` of the commands.

Colfenor replied:
> Not sure what you mean by `have the parser variable duplicated`, but both commands would access the parser variable in their `execute` function for saving the database. I think this is inevitable. I don't see a problem with both the execute functions and the `createDatabaseFromOptions` handling the `parser` variable.

Yes exactly that's what I mean, ok then we have it like that.
I just committed and hopefully the changes are still in the scope of being readable. 
The create & import options man page is also merged now and the import tests are updated with decryption time.

Thanks a lot for your patience & review so far :) !

louib commented:
@Colfenor looking good! Can you do one last rebase?

Colfenor commented:
> @Colfenor looking good! Can you do one last rebase?

yay ! I'm still a bit new to `git rebasing`, do you want me to rebase this branch on top of `develop` ?

droidmonkey commented:
That'll do it, but make sure everything looks good prior to force pushing it back up

Colfenor commented:
> That'll do it, but make sure everything looks good prior to force pushing it back up

Ok thanks for clarification ! I hope it worked out ?

louib commented on src/cli/Utils.cpp:
```diff
@@ -364,4 +364,37 @@ namespace Utils
+    /**
+     * Load a key file from disk. When the path specified does not exist a
+     * new file will be generated. No folders will be generated so the parent
+     * folder of the specified file nees to exist
```
needs**

Colfenor replied:
oh my the obligatory typo :P fixed it !

louib reviewed (approved):
Awesome work @Colfenor , thx a lot for your contribution :tada: :champagne:

louib commented:
@Colfenor tests are failing though. Are they passing locally?

Colfenor commented:
> 
> 
> @Colfenor Ubuntu tests are failing though. Are they passing locally?

@louib I checked the TeamCity build log (via guest login) and apparently the `TestPassphraseGenerator.cpp` -Test failed, 
I saw this randomly fail on other commits I've made and assume this generates in rare cases combinations which are not condsidered by the expected test matches and therefore it fails ? 

I haven't touched this test in this PR.

droidmonkey commented:
DONE! Great work 🎊

Colfenor commented:
Thank you :) and thanks a lot for reviewing & support !
